File: assign4/scala/src/main/scala/RodCutter.scala
object RodCutter {

  def getBestSolution(size : Int, prices : Map[Int, Int]): (Int, List[List[Int]]) = size match{
    case 0 => (0, List(List()))

    case _ => {
      val initialSolution = (pickPrice(size, prices), List(List(size)))

      (1 to size - 1).foldLeft(initialSolution){
        (result, e) => findBestSolution(result, getBestSolutionsForSplit( e , size, prices))
      }
    }
  }

  def getBestSolutionsForSplit(split: Int, size: Int, prices: Map[Int, Int]): (Int, List[List[Int]]) = {
    val firstList = getBestSolution(split, prices)
    val secondList = getBestSolution(size - split, prices)
    val combinedPrice = firstList._1 + secondList._1
    val combinedList = (for(firstCut <- firstList._2; secondCut <- secondList._2) yield (firstCut ::: secondCut).sorted).distinct
    
    (combinedPrice, combinedList)
  }

  def findBestSolution(firstSolution: (Int, List[List[Int]]), secondSolution: (Int, List[List[Int]])) : (Int, List[List[Int]]) = (firstSolution._1 compare secondSolution._1).signum match  {
    case  1 => firstSolution
    case -1 => secondSolution
    case  0 => {
      val mergedList = (firstSolution._2 ::: secondSolution._2)

      (firstSolution._1, mergedList)
    }
  }

  def pickPrice(size : Int, prices : Map[Int, Int]) = {
    if(prices.contains(size)) prices(size) else 0
  }
}

File: assign4/ruby/src/rod_cutter.rb
class RodCutter
  def getBestSolutions(size, prices)
    if(size == 0)
      return 0, [[]]
    end

    initialSolution = [getPrice(size, prices), [[size]]]
    (1..size - 1).to_a.inject(initialSolution) do |result, split|
      findBestSolution(result, getBestSolutionForSplit(split, size, prices))
    end
  end

  def getBestSolutionForSplit(split, size, prices)
    splitSolution = getBestSolutions(split, prices)
    leftOverSolution = getBestSolutions(size - split, prices)
    combinedPrice = splitSolution[0] + leftOverSolution[0]
    combinedSolution = splitSolution[1].flat_map{|firstCut| leftOverSolution[1].map{|secondCut| (firstCut + secondCut).sort}}

    [combinedPrice, combinedSolution]
  end

  def findBestSolution(firstSolution, secondSolution)
    case firstSolution[0] == secondSolution[0]
      when false
        firstSolution[0] > secondSolution[0] ? firstSolution : secondSolution
      when true
        [firstSolution[0], (firstSolution[1] + secondSolution[1]).uniq]
    end
  end

  def getPrice(size, prices)
    prices.has_key?(size) ? prices[size] : 0
  end
end